pkg/gardenlet/controller/shoot/shoot_control_prepare_migration.go

```diff
@@ -0,0 +1,313 @@
+			return retryutils.UntilTimeout(context.TODO(), 10*time.Second, 10*time.Minute, func(context.Context) (done bool, err error) {
+				botanist, err = botanistpkg.New(o)
+				if err != nil {
+					return retryutils.MinorError(err)
+				}
+				return retryutils.Ok()
+			})
+		}),
+		utilerrors.ToExecute("Retrieve the Shoot namespace in the Seed cluster", func() error {
+			if err := botanist.K8sSeedClient.Client().Get(context.TODO(), client.ObjectKey{Name: o.Shoot.SeedNamespace}, namespace); err != nil {
+				if apierrors.IsNotFound(err) {
+					o.Logger.Infof("Did not find '%s' namespace in the Seed cluster - nothing to be done", o.Shoot.SeedNamespace)
+					return utilerrors.Cancel()
+				}
+			}
+			return nil
+		}),
+	)
+
+	if err != nil {
+		if utilerrors.WasCanceled(err) {
+			return nil
+		}
+		return gardencorev1beta1helper.NewWrappedLastErrors(gardencorev1beta1helper.FormatLastErrDescription(err), err)
+	}
+
+	var (
+		nonTerminatingNamespace = namespace.Status.Phase != corev1.NamespaceTerminating
+		defaultTimeout          = 10 * time.Minute
+		defaultInterval         = 5 * time.Second
+
+		g = flow.NewGraph("Shoot's control plane preparation for migration")
+
+		deploySecrets = g.Add(flow.Task{
+			Name: "Deploying Shoot certificates / keys",
+			Fn:   flow.TaskFn(botanist.DeploySecrets).DoIf(nonTerminatingNamespace),
+		})
+		deployETCD = g.Add(flow.Task{
+			Name:         "Deploying main and events etcd",
+			Fn:           flow.TaskFn(botanist.DeployETCD).RetryUntilTimeout(defaultInterval, defaultTimeout).DoIf(nonTerminatingNamespace),
+			Dependencies: flow.NewTaskIDs(deploySecrets),
+		})
+		scaleETCDToOne = g.Add(flow.Task{
+			Name:         "Scaling etcd up",
+			Fn:           flow.TaskFn(botanist.ScaleETCDToOne).RetryUntilTimeout(defaultInterval, defaultTimeout).DoIf(nonTerminatingNamespace && shootHibernated),
+			Dependencies: flow.NewTaskIDs(deployETCD),
+		})
+		waitUntilEtcdReady = g.Add(flow.Task{
+			Name:         "Waiting until main and event etcd report readiness",
+			Fn:           flow.TaskFn(botanist.WaitUntilEtcdReady).DoIf(nonTerminatingNamespace),
+			Dependencies: flow.NewTaskIDs(deployETCD, scaleETCDToOne),
+		})
+		ensureResourceManagerScaledUp = g.Add(flow.Task{
+			Name:         "ensures that the gardener resource manager is scaled to 1",
+			Fn:           flow.TaskFn(botanist.ScaleGardenerResourceManagerToOne).DoIf(nonTerminatingNamespace),
+			Dependencies: flow.NewTaskIDs(),
+		})
+		keepManagedResourcesObjects = g.Add(flow.Task{
+			Name:         "Keep managed resources objects",
+			Fn:           flow.TaskFn(botanist.KeepManagedResourcesObjects),
+			Dependencies: flow.NewTaskIDs(ensureResourceManagerScaledUp),
+		})
+		deleteAllManagedResources = g.Add(flow.Task{
+			Name:         "Deletes all Managed Resources from the Shoot's namespace",
+			Fn:           flow.TaskFn(botanist.DeleteAllManagedResourcesObjects),
+			Dependencies: flow.NewTaskIDs(keepManagedResourcesObjects),
+		})
```

Comment:
Confusing names task/function names: First keep, and then immediately delete? Why?